prow/crier/controller_test.go

```diff
@@ -0,0 +1,231 @@
+}
+
+func (f triggerReporter) GetName() string {
+	return "triggerReporter"
+}
+
+func (f triggerReporter) ShouldReport(pj *apiv1.ProwJob) bool {
+	return pj.Status.State == apiv1.TriggeredState
+}
+
+//Fake Informer
+//Sets: The Prow Job Test Cases
+//Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	eventChannel chan string
+}
+
+func (f fakeInformer) Get(name string) (*apiv1.ProwJob, error) {
+	var pj *apiv1.ProwJob
+	switch name {
+	case "TestJob1":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.FailureState,
+			},
+		}
+	case "TestJob2":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.TriggeredState,
+			},
+		}
+	}
+
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) listerv1.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() listerv1.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Run"
+	}
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                           { return nil }
+func (fakeInformer) GetController() cache.Controller                                 { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                 { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                       { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                       { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*apiv1.ProwJob, err error) { return nil, nil }
+
+func TestController_processNextItem(t *testing.T) {
+
+	cs := fake.NewSimpleClientset()
+
+	q := kube.RateLimiter("CrierTest")
+	q.Add("TestJob1")
+	q.Add("TestJob2")
+
+	inf := fakeInformer{}
+
+	rp := triggerReporter{}
+	isReported := false
+	rp.reportedItem = &isReported
+
+	nmwrk := 1
+	wg := &sync.WaitGroup{}
+	c := NewController(cs, q, inf, rp, nmwrk, wg)
+
+	c.processNextItem()
+
+	if c.queue.Len() != 1 {
+		t.Errorf("%d messages were on queue; expecting 1", c.queue.Len())
+	}
+	if isReported {
+		t.Error("Prow Job was incorrectly reported")
+	}
+
+	c.processNextItem()
+
+	if c.queue.Len() != 0 {
+		t.Errorf("%d messages were on queue; expecting 0", c.queue.Len())
+	}
+	if !isReported {
+		t.Error("Failed to report Prow Job")
+	}
```

Comment:
might also worth assert any k8s operations the controller made, like https://github.com/kubernetes/sample-controller/blob/master/controller_test.go#L133-L161, but feel free to do that as a follow up